java/org/apache/catalina/realm/GenericPrincipal.java

```diff
@@ -304,10 +333,165 @@ public void logout() throws Exception {
+    }
+
+
+    @Override
+    public boolean isAttributesCaseIgnored() {
+        return (attributes instanceof CaseInsensitiveKeyMap<?>);
+    }
+
+
+    /**
+     * Creates and returns a deep copy of the specified object. Deep-copying
+     * works only for objects of a couple of <i>hard coded</i> types or, if the
+     * object implements <code>java.io.Serializable</code>. In all other cases,
+     * this method returns <code>null</code>.
+     * 
+     * @param obj the object to copy
+     * 
+     * @return a deep copied clone of the specified object, or <code>null</code>
+     *         if deep-copying was not possible
+     */
+    private Object copyObject(Object obj) {
+
+        // first, try some commonly used object types
+        if (obj instanceof String) {
+            return new String((String) obj);
+
+        } else if (obj instanceof Integer) {
+            return Integer.valueOf((int) obj);
+
+        } else if (obj instanceof Long) {
+            return Long.valueOf((long) obj);
+
+        } else if (obj instanceof Boolean) {
+            return Boolean.valueOf((boolean) obj);
+
+        } else if (obj instanceof Double) {
+            return Double.valueOf((double) obj);
+
+        } else if (obj instanceof Float) {
+            return Float.valueOf((float) obj);
+
+        } else if (obj instanceof Character) {
+            return Character.valueOf((char) obj);
+
+        } else if (obj instanceof Byte) {
+            return Byte.valueOf((byte) obj); 
+
+        } else if (obj instanceof Short) {
+            return Short.valueOf((short) obj);
+
+        } else if (obj instanceof BigDecimal) {
+            return new BigDecimal(((BigDecimal) obj).toString());
+
+        } else if (obj instanceof BigInteger) {
+            return new BigInteger(((BigInteger) obj).toString());
+
+        }
+
+        // Date and JDBC date and time types
+        else if (obj instanceof java.sql.Date) {
+            return ((java.sql.Date) obj).clone();
+
+        } else if (obj instanceof java.sql.Time) {
+            return ((java.sql.Time) obj).clone();
+
+        } else if (obj instanceof java.sql.Timestamp) {
+            return ((java.sql.Timestamp) obj).clone();
+
+        } else if (obj instanceof Date) {
+            return ((Date) obj).clone();
+
+        }
+
+        // these types may come up as well
+        else if (obj instanceof URI) {
+            try {
+                return new URI(((URI) obj).toString());
+            } catch (URISyntaxException e) {
+                // not expected
+            }
+        } else if (obj instanceof URL) {
+            try {
+                return new URI(((URL) obj).toString());
+            } catch (URISyntaxException e) {
+                // not expected
+            }
+        } else if (obj instanceof UUID) {
+            return new UUID(((UUID) obj).getMostSignificantBits(),
+                    ((UUID) obj).getLeastSignificantBits());
+
+        }
```

Comment:
Why?
Most of the are immutable anyway. Why not check for `Cloneable`?

Reply:
I was discussing that with Mark and Chris. Both are sure that we need to go with _defensive copies_. Chris was suggesting, that `Cloneable` is not the best choice. Also, with reflection (and also with JNI), it is still possible to change the value of a final field, so there are no really immutable objects.

Reply:
Hmm...hard nut to crack. At the end the developer decides which attributes are required to expose to this map. Maybe I am lacking the expertise, but using reflection is a quite constructed case given that Java 16+ is getting stricter. I still think that one does not need copies of immutable objects because what you at most can modify is the in-memory representation of an attribute, but never in the user store. I might be wrong. Gürtel und Hosenträger!

Reply:
I also mentioned that in this PR's initial comment under _Defensive copies of attribute values_. So, consider my solution as a starting point (e. g. for a in-deep discussion about that). There was not yet any agreement for a strategy so, I started off with a working solution that should be safe. Now we see, that it might even by over-engineered or paranoid.

When modifying the final 'value' field in an `Integer`, I'm quite sure that Java uses that new value subsequently. So, It's not only the in-memory representation.

(Das Sprichwort kannte ich bisher nicht. Sehr treffend :)